Speed up SELinux volume relabeling using mounts MVP

jsafrane opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This is minimal PR that implements SELinux mounting with `-o context=XYZ`.

* Works for iSCSI and CSI volumes only.
* I did not update existing unit tests (yet).
* Hoping it passes CI (with the feature gate disabled).

~WIP:~
* ~Add and fix unit tests~
* ~Update after https://github.com/kubernetes/kubernetes/pull/108180 merges~
* ~Add metrics to track pods that could get broken when all access modes would be used (instead of ReadWriteOncePod only).~

This PR adds SELinux tracking to the VolumeManager's ActualStateOfWorld / DesiredStateOfWorld and acts on it in the reconciler. When the alpha feature gate is enabled, all iSCSI and CSI volumes will be tracked, regardless of their access mode, only to emit metrics when a volume would need two different SELinux context when the feature is extended to all access types.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
Add SELinux context tracking to volume manager

Both ActualStateOfWorld and DesiredStateOfWorld must track SELinux context
of volume mounts.

jsafrane committed:
Add SELinux mount option to NewMounter() and MountDevice()

Let volume plugins decide if they want to mount volumes with "-o
context=XYZ" or let the container runtime relabel the volume on container
startup.

Using NewMounter, as it's the call where a volume plugin gets the other MountOptions.

jsafrane committed:
Implement mounting with -o context= in iSCSI volume plugin

jsafrane committed:
Add SupportsSELinuxContextMount

Add a new call to VolumePlugin interface and change all its
implementations.

Kubelet's VolumeManager will be interested whether a volume supports
mounting with -o conext=XYZ or not to hanle SetUp() / MountDevice()
accordingly.

jsafrane committed:
Add SELinuxMountReadWriteOncePod feature gate

jsafrane commented:
/kind feature
/sig storage

jsafrane committed:
Add CSIDriverSpec.SELinuxMount

The new field tells Kubernetes if the CSI driver supports mounting of
volumes with -o context=XYZ or not.

jsafrane committed:
Add metrics for SELinux context mount

Add separate _errors and _warnings to capture volumes that were rejected
from those will be rejected when the feature is expanded to all access
mode.

jsafrane committed:
Fix existing unit tests

jsafrane committed:
Add SELinux mount support to CSI driver

With some minor refactoring to use common getCSIDriver function.

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -547,6 +563,17 @@ func (asw *actualStateOfWorld) AddPodToVolume(markVolumeOpts operationexecutor.M
 		podObj.mounter = mounter
 	}
 	asw.attachedVolumes[volumeName].mountedPods[podName] = podObj
+	if utilfeature.DefaultFeatureGate.Enabled(features.ExpandInUsePersistentVolumes) {
```
This is using `ExpandInUsePersistentVolumes` feature gate.

jsafrane replied:
Fixed

gnufied commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -704,6 +737,22 @@ func (asw *actualStateOfWorld) PodExistsInVolume(
 		return false, "", newVolumeNotAttachedError(volumeName)
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.ExpandInUsePersistentVolumes) {
```
same as above.

jsafrane replied:
Fixed

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -259,7 +274,62 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+	var isRWOP bool
+
+	if feature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) {
+		pluginSupportsSELinuxContextMount = dsw.getSELinuxMountSupport(volumeSpec)
```
CSI support appears to be `false` here?

jsafrane replied:
It should be taken from CSIDriver now

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -259,7 +274,62 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+					fullErr := fmt.Errorf("failed to construct SELinux label from context %q: %s", containerContext, err)
+					if isRWOP {
+						// Cannot mount with -o context if the context can't be composed.
+						seLinuxContainerContextErrors.Add(1.0)
```
Does it make sense for these metrics to count volume type?

jsafrane replied:
Yes, I will add plugin name + the "highest" access mode (a PV with rwx + rwo would get a metric for rwx)

jsafrane replied:
Added a TODO item to desired_state_of_wold_selinux_metrics.go

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -259,7 +274,62 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+	// that would break if the feature was for all volume access modes.
+	var isRWOP bool
+
+	if feature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) {
```
nit: I would perhaps move this whole thing to a function of its own.

gnufied commented on pkg/kubelet/volumemanager/reconciler/reconciler.go:
```diff
@@ -198,9 +198,14 @@ func (rc *reconciler) unmountVolumes() {
+		volMounted, devicePath, err := rc.actualStateOfWorld.PodExistsInVolume(volumeToMount.PodName, volumeToMount.VolumeName, volumeToMount.SELinuxLabel)
 		volumeToMount.DevicePath = devicePath
-		if cache.IsVolumeNotAttachedError(err) {
+		if cache.IsSELinuxMountMismatchError(err) {
```
Shouldn't we modify `reconstruct_new.go` file from reconstruction PR, rather than here? I want to see how it fits together.

jsafrane replied:
I removed all reconstruction changes from this PR, indeed it will need far more changes than presented here.

gnufied commented on pkg/volume/util/util.go:
```diff
@@ -597,15 +599,29 @@ func IsLocalEphemeralVolume(volume v1.Volume) bool {
 
 	podutil.VisitContainers(&pod.Spec, podutil.AllFeatureEnabledContainers(), func(container *v1.Container, containerType podutil.ContainerType) bool {
+		var seLinuxOptions *v1.SELinuxOptions
+		if utilfeature.DefaultFeatureGate.Enabled(features.ExpandInUsePersistentVolumes) {
```
wrong feature gate.

jsafrane replied:
Fixed

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -366,6 +466,15 @@ func (dsw *desiredStateOfWorld) PodExistsInVolume(
 		return false
 	}
 
+	if utilfeature.DefaultFeatureGate.Enabled(features.ExpandInUsePersistentVolumes) {
```
wrong feature gate btw.

jsafrane replied:
Fixed

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -366,6 +466,15 @@ func (dsw *desiredStateOfWorld) PodExistsInVolume(
+			// with the right SELinux option.
+			return false
+		}
+	}
```
When do we expect context is ASOW to be not same as one is desired state? If kubelet was down and another pod got created on same node with different selinux context?

jsafrane replied:
There are two cases:

* As you pointed out, kubelet restart + reconstruction + user created a new pod with a new context.
* User creates two pods with different contexts. The first pod will get to DSW + get mounted + to ASW. The second pod will be rejected by DSW as having a wrong context. Now, when the first pod is deleted, it leaves DSW, the second pod enters DSW, but something must unmount the volume and mount it again with a different context.

gnufied replied:
I wonder if it should be written in such a way that - selinux_context + volume_name,  make for key in ASOW and DSOW maps, because that is our intention at the end of day.

jsafrane replied:
It will be problematic to see if a volume is already mounted with a different SELinux context. Reconciler will try to always mount such a volume and errors from `/bin/mount` about wrong SELinux contexts are ugly (`mount point not mounted or bad option`, without any note about SELinux).

gnufied replied:
Yeah by default, it would be problematic. But for ReadWriteOnce pods, shouldn't another pod not try to mount same volume at the same time? (and it does well - mount error is a valid error).

jsafrane commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -259,7 +274,62 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+	if feature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) {
+		pluginSupportsSELinuxContextMount = dsw.getSELinuxMountSupport(volumeSpec)
+		isRWOP = util.IsRWOP(volumeSpec)
+		if pluginSupportsSELinuxContextMount {
```
Note to self: move the whole selinux + metric to a separate function

jsafrane replied:
Done

logicalhan commented:
/assign @dgrisonnet

jsafrane committed:
Update generation when SELinuxMount is changed

jsafrane committed:
Regenerate files

jsafrane committed:
Extract getSELinuxLabel from AddPodToVolume

To keep the function smaller.

jsafrane commented:
Rebased heavily and removed reconstruction code, a separate PR will be needed for it,

leilajal commented:
/remove-sig api-machinery

msau42 commented on pkg/volume/csi/csi_mounter.go:
```diff
@@ -245,6 +245,35 @@ func (c *csiMountMgr) SetUpAt(dir string, mounterArgs volume.MounterArgs) error
 		}
 	}
 
+	// Save volume info in pod dir
```
Is this just a cherry pick of https://github.com/kubernetes/kubernetes/pull/108874? Should we submit that separately?

reading the original pr that moved this to NewMounter, it sounds like at the time it was done for code cleanliness and not related to fixing the original issue: https://github.com/kubernetes/kubernetes/pull/64519

jsafrane replied:
You're right, I removed this commit.

msau42 commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -365,6 +366,9 @@ func (b *iscsiDiskMounter) SetUpAt(dir string, mounterArgs volume.MounterArgs) e
 	}
+	// The volume must have been mounted in MountDevice with -o context.
+	// TODO: extract from mount table in GetAttributes() to be sure?
+	b.mountedWithSELinuxContext = mounterArgs.SELinuxLabel != ""
```
Does this need the feature gate check? Because even if this is non-nil, the feature gate may be disabled.

jsafrane replied:
IMO it's impossible to get non-empty `mounterArgs.SELinuxLabel` here, but I add the feature gate check to avoid future regressions.

msau42 commented on pkg/kubelet/volumemanager/cache/actual_state_of_world_test.go:
```diff
@@ -590,7 +590,7 @@ func Test_MarkDeviceAsMounted_Positive_NewVolume(t *testing.T) {
 
 	// Act
-	err = asw.MarkDeviceAsMounted(generatedVolumeName, devicePath, deviceMountPath)
+	err = asw.MarkDeviceAsMounted(generatedVolumeName, devicePath, deviceMountPath, "")
```
Can you add a unit test that checks the selinuxcontext is set?

jsafrane replied:
Added a few tests.

msau42 commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -268,7 +279,63 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+				newLabel, err := util.SELinuxOptionsToFileLabel(containerContext)
+				if err != nil {
+					fullErr := fmt.Errorf("failed to construct SELinux label from context %q: %s", containerContext, err)
+					if isRWOP {
```
This error processing is repeated a few times. Maybe worth putting into a function, especially if you plan to add metrics later

jsafrane replied:
Extracted into a function.

msau42 commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -380,16 +463,22 @@ func (dsw *desiredStateOfWorld) UpdatePersistentVolumeSize(volumeName v1.UniqueV
+		return false
+	}
+	if feature.DefaultFeatureGate.Enabled(features.SELinuxMountReadWriteOncePod) {
+		return vol.seLinuxFileLabel == seLinuxMountContext
```
why does selinux matching impact volume existance?

jsafrane replied:
Because a volume with the same UniqueVolumeName, but a different SELinux label is *a different volume* from Kubelet point of view. Kubelet must unmount the it and mount it again with the right `-o context`.

I tried to add the SELinux context to `UniqueVolumeName`, but it is used by A/D controller and `node.status.volumesInUse` / `volumesAttached` and here we don't care about SELinux.

I tried to use another key for kubelet's ASW / DSW, but it got very messy quickly.

jsafrane replied:
Added a long comment

msau42 commented on pkg/kubelet/volumemanager/cache/desired_state_of_world_test.go:
```diff
@@ -59,7 +60,7 @@ func Test_AddPodToVolume_Positive_NewPodNewVolume(t *testing.T) {
 	// Act
 	generatedVolumeName, err := dsw.AddPodToVolume(
-		podName, pod, volumeSpec, volumeSpec.Name(), "" /* volumeGidValue */)
+		podName, pod, volumeSpec, volumeSpec.Name(), "" /* volumeGidValue */, nil /* seLinuxContainerContexts */)
```
add a unit test using selinux context

jsafrane replied:
Added unit tests for `AddPodToVolume`. I have some doubts about go-selinux package, it seems it actually needs SELinux enabled during the unit tests.

TODO: add some interface between kubelet and go-selinux.

jsafrane replied:
Added an interface for SELinux label translation.

jsafrane committed:
Extract SELinux context error handling into a common func

Add handlerSELinuxMetricError() which bumps the right metric + either
consumes a SELinux error or lets it propagate up the stack.

jsafrane committed:
Add SELinux feature check for iSCSI volume plugin

In theory the check is not necessary, but for sake of robustness and
completenes, let's check SELinuxMountReadWriteOncePod feature gate before
assuming anything about SELinux labels.

jsafrane commented:
I fixed everything except for requested new unit tests, working on them right now.

jsafrane committed:
Add AddPodToVolume unit tests with SELinux

jsafrane committed:
Add interface for SELinuxOptionsToFileLabel

github.com/opencontainers/selinux/go-selinux needs OS that supports SELinux
and SELinux enabled in it to return useful data, therefore add an interface
in front of it, so we can mock its behavior in unit tests.

leilajal commented:
/remove-sig api-machinery

jsafrane committed:
Add unit tests for ASW.AddPodToVolume

jsafrane commented:
All required tests added (+few extra ones too).

msau42 commented on pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go:
```diff
@@ -304,7 +304,7 @@ func (dswp *desiredStateOfWorldPopulator) processPodVolumes(
 		// Add volume to desired state of world
 		uniqueVolumeName, err := dswp.desiredStateOfWorld.AddPodToVolume(
-			uniquePodName, pod, volumeSpec, podVolume.Name, volumeGidValue)
+			uniquePodName, pod, volumeSpec, podVolume.Name, volumeGidValue, seLinuxContainerContexts[podVolume.Name])
```
unit test?

jsafrane replied:
Unit tests added

jsafrane committed:
Remove noisy log

The error would be logged every reconciler sync (100 ms).

jsafrane commented on pkg/kubelet/volumemanager/cache/actual_state_of_world.go:
```diff
@@ -785,6 +817,22 @@ func (asw *actualStateOfWorld) PodExistsInVolume(podName volumetypes.UniquePodNa
+				fullErr := newSELinuxMountMismatchError(volumeName)
+				if util.IsRWOP(volumeObj.spec) {
+					return false, volumeObj.devicePath, fullErr
+				} else {
```
There is no metric to bump. And if there was one, it would be incremented every resync interval (100 ms), so it would not be that useful.
I am actually removing the `else` part with log, it would be logged really often and there already is one in `reconciler.mountOrAttachVolumes`.

jsafrane committed:
Rename IsRWOP

To be able to update content of the function to other access modes when we
implement SELinux mount for more of them.

jsafrane commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -283,6 +307,13 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
 				}
 			}
 		}
+		if !util.IsRWOP(volumeSpec) {
```
Renamed to VolumeSupportsSELinuxMount.

jsafrane committed:
Add a coment about handling same volumes with different contexts

jsafrane committed:
Add unit tests for GetPodVolumeNames

jsafrane committed:
Add unit tests for DesiredStateOfWorldPopulator

jsafrane commented:
I am actually removing WIP - we can either merge this PR (API + implementation) or #108744 first and then rebase.

jsafrane commented:
Unit test passed \o/

msau42 commented:
/lgtm
/label api-review

msau42 commented:
/assign @thockin 
for API review.

The first 4 commits are for the API, which was previously reviewed in https://github.com/kubernetes/kubernetes/pull/108744

fedebongio commented:
/remove-sig api-machinery

Priyankasaggu11929 commented:
Hello :wave:, 1.25 Enhancements Lead here.

Unfortunately, this enhancement did not meet the [code freeze](https://github.com/kubernetes/sig-release/blob/master/releases/release_phases.md#code-freeze) criteria because there are still unmerged k/k code PRs.

If you still wish to progress this enhancement in v1.25, please file an [exception request](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md). Thank you so much!

cici37 commented:
Hello 👋, 1.25 Release Lead here.

The [exception request](https://groups.google.com/a/kubernetes.io/g/release-team/c/r1PMmU6koyc) is approved and your updated deadline to make any changes to your PR is 3:30 AM PST Friday 5th August 2022. Thank you!

/milestone v1.25

thockin commented:
API LGTM (carryover, but I did go look at it :)
/approve

jsafrane commented:
/priority important-soon

jsafrane commented:
/retest

This failed:
> Services should fail health check node port if there are only terminating endpoints

RomanBednar commented:
/lgtm

jsafrane commented:
ipv6 tests failed with: Pods Extended Pod Container lifecycle should not create extra sandbox if all containers are done
/retest

jsafrane commented:
/test pull-kubernetes-dependencies
The test [passed](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108692/pull-kubernetes-dependencies/1555178940534886400), but it's still yellow here for couple of hours.

cici37 commented:
/remove-sig api-machinery
